### ISSUE TITLE
Shadowing names in cases of a match with non-trivial scrutinee

### ISSUE THREAD

**Issue by lambda-abstraction** [microsoft/pyright#10632]
**Code**
```python
from typing import Literal
from collections.abc import Mapping

type Ty = (
    tuple[Literal["var"], str] |
    str
)

type Sub = Mapping[str, Ty]

def unify(a: Ty, b: Ty) -> Sub:
    match a, b:
        case ("var", a), b:
            # a is narrowed to tuple[Literal["var", str]] when its really a `str` (both making this well-typed program not typecheck, and allowing absurd programs)
            return {a: b}
        case _:
            assert a == b
            return {}

def trivial_scrutinee(a: Ty) -> str:
    match a:
        case ("var", a):
            # shadowed correctly
            return a
        case _:
            raise ValueError

def nontrivial_scrutinee(a: Ty) -> str:
    match a, None:
        case ("var", a), _:
            # a: tuple[Literal["var"], str] so it doesnt typecheck
            return a
        case _:
            raise ValueError
```

**Expected behavior and questions**
In `unify` i expected the type of `a` to get shadowed to `str`, but i guess that python typecheckers usually dont like mid-scope shadowing?
But in this specific case, the `case` is "closed" by a return statement, so it does not affect the type of `a` in code after the match, so i'd expect it to be fine
What's pyright's stance on shadowing in general? Is it documented somewhere? I couldn't find it by ctrl-f'ing "shadowing" in https://microsoft.github.io/pyright/#/type-concepts-advanced

**VS Code extension or command-line**
VSCode and CLI, pyright 1.1.402

**Comment by erictraut** [comment]
Thanks for reporting. I agree this is a bug.

> What's pyright's stance on shadowing in general?

The term "shadowing" refers to the situation where an inner-scoped variable has the same name as an outer-scoped variable. In your examples, there is only one scope involved, so there is no "shadowing" here. There is only one symbol `a` in each of your example functions, so that's how pyright models it.

While shadowing isn't occurring here, there is an assignment that happens when a `case` statement matches the pattern. A new value is assigned to `a`, but this occurs after `a`'s pre-assignment type is narrowed. I'm guessing that pyright's implementation has these reversed, which means that the new assignment is being applied, and then the narrowed type of the pre-assigned value is effectively overwriting the newly-assigned type. I'll need to investigate further to validate my hypothesis.

In the meantime, you can work around the issue by using a variable with a different name.
